apis/v1beta1/gateway_types.go

```diff
@@ -120,18 +104,59 @@ type GatewaySpec struct {
 	// 3. A mixture of "TCP" and "UDP" Protocol Listeners, where no Listener
 	//    with the same Protocol has the same Port value.
 	//
-	// An implementation that cannot serve both TCP and UDP listeners on the same
-	// address, or cannot mix HTTPS and generic TLS listens on the same port
-	// would not consider those cases compatible.
+	// Some fields in the Listener struct have possible values that affect
+	// whether the Listener is distinct. Hostname is particularly relevant
+	// for HTTP or HTTPS protocols.
+	//
+	// When using the Hostname value to select between same-Port, same-Protocol
+	// Listeners, the Hostname value must be different on each Listener for the
+	// Listener to be distinct.
+	//
+	// When the Listeners are distinct based on Hostname, inbound request
+	// hostnames MUST match from the most specific to least specific Hostname
+	// values to choose the correct Listener and its associated set of Routes.
 	//
-	// Implementations using the Hostname value to select between same-Port
-	// Listeners MUST match inbound request hostnames from the most specific
-	// to least specific Hostname values to find the correct set of Routes.
 	// Exact matches must be processed before wildcard matches, and wildcard
 	// matches must be processed before fallback (empty Hostname value)
 	// matches. For example, `"foo.example.com"` takes precedence over
 	// `"*.example.com"`, and `"*.example.com"` takes precedence over `""`.
 	//
+	// Additionally, if there are multiple wildcard entries, more specific
+	// wildcard entries must be processed before less specific wildcard entries.
+	// For example, `"*.foo.example.com"` takes precedence over `"*.example.com"`.
+	// The precise definition here is that the higher the number of dots in the
+	// hostname to the right of the wildcard character, the higher the precedence.
+	//
+	// The wildcard character will match any number of characters _and dots_ to
+	// the left, however, so `"*.example.com"` will match both
+	// `"foo.bar.example.com"` _and_ `"bar.example.com"`.
+	//
+	// If a set of Listeners contains Listeners that are not distinct, then those
+	// Listeners are Conflicted, and the implementation MUST set the "Conflicted"
+	// condition in the Listener Status to "True".
+	//
+	// Implementations MAY choose to accept a Gateway with Conflicted
+	// Listeners if they accept a partial Listener set that contains no
+	// Conflicted Listeners. They MUST set a "ListenersNotValid" condition
```

Comment:
This allows breaking changes to only affect the things that are broken rather than the entire Gateway - I thought that this would be desirable because of that.

Reply:
I think we're saying the same thing here. Basically I'm just trying to make it clear that implementations should only accept conflicted listeners if the Gateway has already been accepted. (I think you're basically saying the same thing in your follow up, but let me know if I'm misunderstanding).
```suggestion
	// If a Gateway has already been Accepted and a Listener becomes invalid
	// or conflicted, implementations MAY choose to support the subset of Listeners
	// that are still valid and distinct. They MUST set a "ListenersNotValid" condition
```

Reply:
I don't like the idea of the spec making statements that involve the implementation holding state. _Practically_, the implementation will hold state, but the more we can have the decisions about Accepted not require referencing the state, the better off we'll be.

To put this another way, why should the "Already accepted" state be treated differently than the "new Gateway" state? All this is saying is, if you accept any part of a set of Listeners, it has to have no conflicts. Rereading what I have, I can see that I haven't said that very clearly though. Let me update and see what you think.

Reply:
To clarify, I was actually meaning if the Gateway status already had "Accepted" set to true. Agree that we should not require implementations to hold state.